examples/profile/profile.py

```diff
@@ -0,0 +1,157 @@
+                    'measurement. This iteration is not included in mean elapsed time.')
+parser.add_argument('--batchsize', '-b', type=int, default=None,
+                    help='batchsize. If None, '
+                    'batchsize is architecture-specific batchsize is used.')
+parser.add_argument('--out-file', '-o', type=str, default=None,
+                    help='path to output file')
+parser.add_argument('--column-prefix', '-P', type=str, default='',
+                    help='prefix of the name of columns of output matrix')
+args = parser.parse_args()
+
+numpy.random.seed(args.seed)
+if args.gpu >= 0:
+    cuda.cupy.random.seed(args.seed)
+
+
+# model setup
+if args.predictor == 'alex':
+    predictor = alex.Alex(args.batchsize, args.cudnn)
+elif args.predictor == 'overfeat':
+    predictor = overfeat.Overfeat(args.batchsize, args.cudnn)
+elif args.predictor == 'vgg':
+    predictor = vgg.VGG(args.batchsize, args.cudnn)
+elif args.predictor.startswith('conv'):
+    number = args.predictor[4:]
+    predictor = getattr(conv, 'Conv{}'.format(number))(args.batchsize, args.cudnn)
+else:
+    raise ValueError('Invalid predictor name')
+
+model = links.Classifier(predictor)
+
+if args.gpu >= 0:
+    cuda.get_device(args.gpu).use()
+    model.to_gpu()
+xp = cuda.cupy if args.gpu >= 0 else numpy
+
+optimizer = optimizers.SGD()
+optimizer.setup(model)
+
+
+print('Architecture\t{}'.format(args.predictor))
+print('Iteration\t{}'.format(args.iteration))
+print('Cache\t{}'.format(args.cache_level))
+print('cuDNN\t{}'.format(cuda.cudnn_enabled and model.predictor.use_cudnn))
+print('Batchsize\t{}'.format(model.predictor.batchsize))
+
+
+def clear_cache(cache_level):
+    if args.gpu < 0:
+        return
+    if cache_level is None:
+        cache_dir = cuda.cupy.cuda.get_cache_dir()
+        if os.path.exists(cache_dir):
+            shutil.rmtree(cache_dir)
+    if cache_level is None or cache_level == 'disk':
+        cuda.cupy.clear_memo()
+
+
+if args.cache_level is None:
+    start_iteration = 0
+else:
+    start_iteration = -1
+
+forward_time = 0.0
+backward_time = 0.0
+update_time = 0.0
+
+print('iteration\tforward\tbackward\tupdate (in seconds)')
+for iteration in six.moves.range(start_iteration, args.iteration):
+    clear_cache(args.cache_level)
+
+    # data generation
+    data = numpy.random.uniform(-1, 1,
+                                (model.predictor.batchsize,
+                                 model.predictor.in_channels,
+                                 model.predictor.insize,
+                                 model.predictor.insize)).astype(numpy.float32)
+    data = chainer.Variable(xp.asarray(data))
+    label = numpy.empty((args.batchsize,), dtype=numpy.int32)
+    label.fill(0)
+    label = chainer.Variable(xp.asarray(label))
+
+    # forward
+    with function_hooks.TimerHook() as h:
+        loss = model(data, label)
+    forward_time_one = h.total_time()
+
```

Comment:
OK, I found that the timer hook actually synchronizes with the CUDA streams, so it actually waits for all kernels to finish.